src/auth/mod.ts

```diff
@@ -15,33 +14,131 @@ interface DecodedToken {
+    true,
+    ["verify"],
   );
+  const verified = await verify(token, cryptoKey);
+  return verified as DecodedToken;
 }
 
-export function decodeToken(token: string): DecodedToken | null {
-  if (token.startsWith("Bearer ")) {
-    token = token.slice(7);
+async function getPublicKey(): Promise<string> {
+  const jwks_uri = await getJWKURI();
+  const jwkUrl = new URL(jwks_uri);
+  const headers = new Headers();
+  if (envVarsConfig.env === "DEV") {
+    // This should be the proxy server created by kubectl for dev purposes inside a local machine
+    jwkUrl.hostname = envVarsConfig.k8s_api;
+  } else {
+    // we need to set the auth headers with the serviceToken when herald is running in a k8s cluster
+    headers.set(
+      "Authorization",
+      `Bearer ${await getServiceAccountToken()}`,
+    );
+  }
+  const fetchJWK = await fetch(jwkUrl.toString(), { headers });
+  const data = await fetchJWK.json();
+  const key = data.keys[0].kid;
+  if (!key) {
+    throw new Error("Key not found in response");
   }
-  const [_headers, payload, _signature] = decode(token);
 
-  const data = payload as DecodedToken;
-  if (!data || typeof data !== "object" || !data.sub) {
-    logger.warn("Invalid token");
-    throw new HTTPException(401, {
-      message: "Invalid token",
-    });
+  return key;
+}
+
+async function getJWKURI(): Promise<string> {
+  const env = envVarsConfig.env;
+  if (env === "DEV") {
+    const uri = Deno.env.get("JWKS_URI");
+    if (!uri) {
+      logger.error("JWKS_URI not found in environment");
+      throw new Error("JWKS_URI not found in environment");
+    }
+    return uri;
   }
+  const k8s_url = envVarsConfig.k8s_api;
+  const fetchJWKURI = await fetch(
+    `${k8s_url}/.well-known/openid-configuration`,
+    {
+      headers: {
+        Authorization: `Bearer ${await getServiceAccountToken()}`,
+      },
+    },
+  );
+  const data = await fetchJWKURI.json();
+  const jwks_uri = data.jwks_uri;
+
+  if (!jwks_uri) {
+    logger.error("JWKS URI not found in response");
+    throw new Error("JWKS URI not found in response");
+  }
+
+  return jwks_uri;
+}
 
-  if (!isSubInPodsConfig(data)) {
-    logger.warn("Unauthorized");
-    throw new HTTPException(403, {
-      message: "Unauthorized",
+async function getServiceAccountToken(): Promise<string> {
+  const env = envVarsConfig.env;
+  if (env === "DEV") {
+    const podName = Deno.env.get("POD_NAME");
+    if (!podName) {
+      logger.error("POD_NAME not found in environment");
+      throw new Error(
+        "POD_NAME not found in environment. Set the ENV variable to PROD if running in a Kubernetes cluster",
+      );
+    }
+    const namespace = Deno.env.get("NAMESPACE");
+    if (!namespace) {
+      logger.error("NAMESPACE not found in environment");
+      throw new Error(
+        "NAMESPACE not found in environment. Set the ENV variable to PROD if running in a Kubernetes cluster",
+      );
+    }
+    const getTokenCommand = new Deno.Command("kubectl", {
```

Comment:
you will not have kubectl in the container, you should read the file mounted by kubernetes at the expected location (see data cycle)

Reply:
This is for a dev env't

Reply:
Oh I see, let's continue on the other comment ↑